core/cont/inc/TCollection.h

```diff
@@ -296,78 +298,11 @@ namespace Internal {
 const TCollection &EmptyCollection();
 bool ContaineeInheritsFrom(TClass *cl, TClass *base);
 
-/// @brief Internal help class implmenting an iterator for TRangeDynCast.
-template <class Containee> // Containee must derive from TObject.
-class TRangeDynCastIterator : public TIter {
-   static_assert(std::is_base_of<TObject, Containee>::value, "Containee type must inherit from TObject");
-
-   /// This is a workaround against ClassDefInline not supporting classes
-   /// missing their default constructor or having them private.
-   template <class T>
-   friend class ROOT::Internal::ClassDefGenerateInitInstanceLocalInjector;
-
-   TRangeDynCastIterator() = default;
-
-public:
-   using TIter::TIter;
-   TRangeDynCastIterator(const TIter &iter) : TIter(iter) {}
-
-   Containee *operator()() = delete;
-
-   Containee *Next() { return dynamic_cast<Containee *>(TIter::Next()); }
-   Containee *operator*() const { return dynamic_cast<Containee *>(TIter::operator*()); }
-
-   ClassDefInline(TRangeDynCastIterator, 0);
-};
-
 } // namespace Internal
 
 namespace Detail {
 
-/// @brief TTypedIter is a typed version of TIter.
-///
-/// This requires the collection to contains elements of the type requested
-/// (or a derived class).  Any deviation from this expectation
-/// will only be caught/reported by an assert in debug builds.
-///
-/// This is best used with a TClonesArray, for other cases prefered TRangeDynCast.
-///
-/// The typical use is:
-/// ```{.cpp}
-///    TTypedIter<TBaseClass> next(tbaseClassClonesArrayPtr);
-///    while(auto bcl = next()) {
-///       ... use bcl as a TBaseClass*
-///    }
-/// ```
-template <class Containee> // Containee must derive from TObject.
-class TTypedIter : public TIter {
-   static_assert(std::is_base_of<TObject, Containee>::value, "Containee type must inherit from TObject");
-
-   /// This is a workaround against ClassDefInline not supporting classes
-   /// missing their default constructor or having them private.
-   template <class T>
-   friend class ROOT::Internal::ClassDefGenerateInitInstanceLocalInjector;
-
-   TTypedIter() = default;
-
-   static Containee *StaticCast(TObject *obj)
-   {
-      assert(!obj || ROOT::Internal::ContaineeInheritsFrom(obj->IsA(), Containee::Class()));
-      return static_cast<Containee *>(obj);
-   }
-
-public:
-   using TIter::TIter;
-   TTypedIter(const TIter &iter) : TIter(iter) {}
-
-   Containee *operator()() { return StaticCast(TIter::Next()); }
-   Containee *Next() { return StaticCast(TIter::Next()); }
-   Containee *operator*() const { return StaticCast(TIter::operator*()); }
-
-   ClassDefInline(TTypedIter, 0);
-};
-
-/// @brief TRangeStaticCast is an adaptater class that allows the typed iteration
+/// @brief TRangeStaticCast is an adapter function that allows the typed iteration
```

Comment:
`TRangeStaticCast` used to be a class; it's awkward (be it for coding conventions alone) to change that to be a function. Do you think you can keep the class interface? Is there an advantage of making `RangeXYZCast` a function interface instead? I can see a downside, namely the limited temporary lifetime extension.

Reply:
Sure! I brought the class interface back for `TRangeStaticCast` and `TRangeDynCast`, but for the more general `static_range_cast` and `dynamic_range_cast` I could not do it (or now `StaticRangeCast` and `DynamicRangeCast` as you suggested to rename them).

The problem is that these functions have a second template parameter, which is the type of the input container. For classes, you'd need to pass the template argument explicitly which is what I don't want here. Or is there another solution for this?

Reply:
Or I can just put the functions only in RooFit, maybe that's less confusing because I will not introduce new free functions in `core`.

Reply:
These functions are great and we should have them. But it'd be good if you can find a way to keep the `TCollection` versions classes - I suppose we know the iterator type after all?